### PR TITLE
Data loading

### PR DESCRIPTION
**Pull request by SodaVolcano**
---
Title: "Closes issue #22 "
About: Function to load PSV files and tests
Labels: ''
Assignees: ''

---

## **What does this implement/fix? Explain your changes.**
Adds 2 functions to load a directory of PSV files and output a dictionary of `LazyFrame`s
- Removes data sync script

# **Checklist:**
- [x] My code follows the code style of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Review by ctmes — APPROVED** [review]

**Review by Reece030130 — APPROVED** [review]

File: nhs/data/handling.py [hunk 1]
```diff
@@ -13,27 +14,45 @@ def read_psv(file_path: str) -> Optional[pl.LazyFrame]:
     return pl.scan_csv(file_path, separator="|")
 
 
-def read_all_psv(file_dir: str, key_pattern: str) -> dict[str, pl.LazyFrame | None]:
```
Comment (Reece030130) [inline nhs/data/handling.py]: Love the change.

**Comment by Reece030130** [comment]
Census data contains .csv files. which will be filtered when read_all_psv. and xlsx files. Do we need read_all_csv......

**Comment by Reece030130** [comment]
> Census data contains .csv files. which will be filtered when read_all_psv. and xlsx files. Do we need read_all_csv......

change the filter should be good.

**Review by LeyiChen62 — APPROVED** [review]

**Comment by SodaVolcano** [comment]
> Census data contains .csv files. which will be filtered when read_all_psv. and xlsx files. Do we need read_all_csv......

I've added support for `csv` and `xlsx`, function is now called `read_spreadsheets`. I didn't add error checking because that'll muddy up the code and the return type, I'm assuming the user can use the function properly.

File: nhs/__init__.py
Comment (zazwaz12) [inline nhs/__init__.py:1]: Doesn't add any additional functionality - but could be good for preventing errors and missing documentation.

**Review by JessicaShindunata — APPROVED** [review]
Noted, will keep these in mind

File: nhs/data/handling.py
Comment (zazwaz12) [inline nhs/data/handling.py:1]: The docstrings are a bit inconsistent, and missing on __get_spreadsheet_reader - but they are also very simplistic so thats fine as long as the rest of the group agrees. 

Good use of Literal - first time I have seen Callable used; neat.

We need to clarify when we are using "__", is "__main__.py" used differently to "main.py" and is "__get_spreadsheet_reader" the convention for a Callable function?

I appreciate the abstraction of regex to a string formatting module.

File: nhs/utils/logging.py
Comment (zazwaz12) [inline nhs/utils/logging.py:1]: Quite hard to follow, is the Logging with Loguru Tutorial on the Wiki all encompassing? We only ever need to worry about calling @log_entry_exit("INFO"), decorator and I assume we need it for all the module. Could not see where the entry and exit came from.

File: scripts/context.py
Comment (zazwaz12) [inline scripts/context.py:1]: Does this add extra functionality? I thought pytest runs through the whole nhs directory.

File: nhs/utils/string.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        "/path/to/{organ}_{observer}.nii.gz",
+        ["organ", "observer"],
+        r"[^b]+",    # Any string that does not contain the letter 'b'
+    )
+    # Output
+    [("eye", "sp")]
+
+    ```
+    """
+    return tz.pipe(
```
Comment (zazwaz12) [inline nhs/utils/string.py:111]: @SodaVolcano

**Review by zazwaz12 — COMMENTED** [review]
Good to merge, we will get use to using the loguru set-up + the decorators for it and what tz.pipe + curried does.

File: nhs/__init__.py
Comment (SodaVolcano) [reply nhs/__init__.py:1]: oh `__init__.py` does add functionality even if the file is empty. A folder with a `__init__.py` file tells python to treat that folder and the files in it as a *package*, meaning you can use relative imports
```python
from .data import some_func   # reference the function some_func inside ./data.py
```
It would result in an error if the `__init__.py` file was missing. Some past experimentation told me using relative import is probably better for files inside `nhs` :sweat_smile:

File: nhs/data/handling.py
Comment (SodaVolcano) [reply nhs/data/handling.py:1]: For simpler functions a one-line doc-string is fine, otherwise it will make the code more unreadable. The types in the function signature serve as documentation even if we are not using `pyright`, and the code is self-explanatory with the variable names. 

I'm not sure if `__get_spreadsheet_reader` needs documentation given how short it is, and functions beginning with `_` or `__` means it's private (so it shouldn't be used by the end-user). It's purely for internal usage within this file but I'll add it in.

`Callable` are functions, so that includes `read_psv` etc. It's just something that let you specify the input types and output types instead of having a type called `function` with no additional information. The `Callable` in the code refers to the function that `__get_spreadsheet_reader` RETURNS, NOT the `__get_spreadsheet_reader` function itself.

For `__main__.py`, that file is only executed when `nhs` is being run as a **module**, i.e. when we run
```bash
python3 -m nhs
```
it's different from `main.py` (I'm not sure if most libraries have `main.py` and if it's necessary).

File: nhs/utils/string.py
[the remark this answers is not in the document]
Comment (SodaVolcano) [reply nhs/utils/string.py:1]: `toolz.pipe` take in one value as its first argument and a list of functions. If you click inside the function it's literally just a for-loop
```python
def pipe(data, *funcs):
    for func in funcs:
        data = func(data)
    return data
```
so it just passes the first parameter through the rest of the parameters (which are all functions).

`curried` lets you have functions with only some values filled in. Normal python `map(func, iterable)` REQUIRE two arguments, a function `func` and an iterable `iterable` (like a list, tuple, dictionary etc). But if you only have `func` and not the data until later on, doing `map(func)` will give you an error because it's missing the `iterable` parameter. A curried map won't give you this error. `curried.map(func)` will give you a **new function** that you can pass around and fill in with `iterable` when you have the data.

In the code here, `curried` is just used as a short-hand to remove some redundancy in the code. Without `tz.pipe` and `curried`, the code is (using `capture_placeholders` as an example)

```python
 return tz.pipe(
        [s] + placeholders,
        curried.reduce(
            lambda string, placeholder: string.replace("{" + placeholder + "}", "\x00")
        ),
        lambda string: re.sub(r"{[a-zA-Z0-9_]*}", "\x01", string),
        re.escape,
        lambda string: string.replace("\x00", f"({re_pattern})"),
        lambda string: string.replace("\x01", re_pattern),
        str,
    )

x = [s] + placeholders
x = reduce(
            lambda string, placeholder: string.replace("{" + placeholder + "}", "\x00"),
            x
 )
x = re.sub(r"{[a-zA-Z0-9_]*}", "\x01", x)
x = re.escape(x)
x = x.replace("\x00", f"({re_pattern})"),
x = x.replace("\x01", re_pattern),
x = str(x)
```

Honestly the syntax should probably be changed to not use `toolz` since the client don't want to deal with it. I left it here because
1. The code is copied and pasted from another project
2. It's tested and works for expected cases (unit tests were also copied)
3. Rewriting it is gonna take extra time

This can be made into an issue though if you think it should be refactored

File: scripts/context.py
Comment (SodaVolcano) [reply scripts/context.py:1]: This is for the imports. `tests` is in a separate folder from `nhs` and so you can't do `from nhs import ...`

You can see the test files in `tests` all import from `context`
